# Genomic Data Science Specialization - Johns Hopkins University (Coursera)

---

No. | Completed | Course |
--- | --- | --- |
I   | Completed | [Introduction to Genomic Technologies](https://www.coursera.org/learn/introduction-genomics?specialization=genomic-data-science)
II  | Completed | [Python for Genomic Data Science](https://www.coursera.org/learn/python-genomics?specialization=genomic-data-science)
III | Completed | [Algorithms for DNA Sequencing](https://www.coursera.org/learn/dna-sequencing?specialization=genomic-data-science) |
IV  | In Progress | [Command Line Tools for Genomic Data Science](https://www.coursera.org/learn/genomic-tools?specialization=genomic-data-science) |
V   | Queued | [Bioconductor for Genomic Data Science](https://www.coursera.org/learn/bioconductor?specialization=genomic-data-science) |
VI  | Queued | [Statistics for Genomic Data Science](https://www.coursera.org/learn/statistical-genomics?specialization=genomic-data-science) |

---

## I. Introduction to Genomic Technologies
 - Biology Overview
 - Measurement Technology
 - Computing Technology
 - Data Science Technology

## II. Python for Genomic Data Science
- Strings and Numbers
- Data Structures, Ifs, and Loops
- Functions and Modules
- Reading Data
- Biopython

## III. Algorithms for DNA Sequencing
- DNA Sequencing
- Exact and Approximate Matching
- Edit Distance and Alignment
- Read Assembly

## IV. Command Line Tools for Genomic Data Science
- Unix Commands and Pipes
- Sequences and genomic features
  - Formats: Fastq, BED, GTF, SAM
  - Tools: Samtools, Bedtools
- Alignment and sequence variation
  - Bowtie, BWA
  - Samtools, Bcftools
  - Application: variant calling
- Transcriptomics
  - Tophat
  - Cufflinks
  - IGV
  - Application: RNA-seq analysis
 
## V. Bioconductor for Genomic Data Science
- Week 1
   - Introduction to Bioconductor
   - GRanges and AnnotationHub
   - IRanges
- Week 2
   - Biological strings in Bioconductor
   - Biostrings
   - BSgenome
   - Intermediate usage of GenomicRanges
   - Rle
   - Transcription annotation in Bioconductor
   - GenomicFeatures
   - Importing data using rtracklayer
- Week 3
   - Data containers and annotations
   - ExpressionSet
   - SummarizedExperiment
    - GEOquery
    - biomaRt
- Week 4
   - Handling raw and aligned short reads
   - Basic usage of statistical models for differential expression
   - Preprocessing microarrays

## VI. Statistics for Genomic Data Science
- Week 1
   - Reproducible Research
   - The three tables in genomics
   - Experimental Design: Variability, replication, power
   - Experimental Design: Confounding and randomization
   - Exploratory Analysis
   - Data transforms
   - Clustering
- Week 2
   - Dimension reduction
   - Pre-processing and normalization
   - Quantile normalization (in R)
   - Linear models
   - Linear models with categorical covariates
   - Adjusting for covariates
   - Linear regression (in R)
   - Many regressions at once
   - Batch effects and confounders (in R)
- Week 3
   - Logistic regression
   - Regression for counts
   - GLMs (in R)
   - Inference
   - Null and alternative hypotheses
   - Calculating statistics
   - Comparing models
   - Permutation
   - P-values
   - Multiple testing
- Week 4
   - Gene set analysis
   - More enrichment
   - The process for RNA-seq
   - The process for Chip-Seq
   - The process for DNA methylation
   - The process for GWAS/WGS
   - Combining data types (eQTL)
   - Researcher degrees of freedom
   - Inference vs. prediction
